### ISSUE TITLE
Proposal for a `dotOperator` replacement

### ISSUE THREAD

**Issue by n0bra1n3r** [nim-lang/RFCs#372]
## Description

* The result of a defined "type-bound operator" is called when a call to an undeclared routine is encountered.
* Effectively replaces the experimental [`dotOperator`](https://nim-lang.org/docs/manual_experimental.html#special-operators-dot-operators) feature.

## Illustration

```nim
type Proxy = object

macro `=call`(obj: Proxy, op: untyped, args: varargs[typed]) =
  echo op.repr
  # implementation
  ...

let proxy = Proxy()

discard proxy.test # prints 'test'
discard proxy.?test # prints '?test'
proxy.test = 1 # prints 'test='
proxy.test += 1 # prints 'test+='
`test*=`(proxy, 1) # prints 'test*='
```

## Rationale

* Very useful for adhoc interop with other languages and remote resources.
* Leverages Nim's existing features such as [UFCS](https://nim-lang.org/docs/manual.html#procedures-method-call-syntax), [command invocation syntax](https://nim-lang.org/docs/manual.html#procedures-command-invocation-syntax), and operators to naturally provide custom syntax.

## Definition of Terms

I use the term "routine" here to mean anything that's callable, including procs, operators, templates, and macros.

## Usecases

**Interop:**
```nim
type CppObj {.importcpp header:"CppObj.h".} = object

template `=call`(obj: CppObj, op: untyped) =
  proc op(obj: CppObj) {.importcpp.}

let cppObj = CppObj()
cppObj.cppMethod()
```
Note that this technique can also be used to call arbitrary functions from DLLs, allowing runtime interop and custom REPL-like functionality.

**Field swizzling (GLSL-like field access):**
```nim
type Vector3 = object
  x, y, z: float
type Vector2 = object
  x, y: float

macro `=call`(vec: Vector3, op: untyped) =
  let obj = nnkObjConstr.newTree(ident("Vector" & $len($op)))
  let fields = vec.getType[2]

  for i, field in $op:
    obj.add(newColonExpr(fields[i], newDotExpr(vec, ident($field))))

  template makeProc(op, obj) =
    proc op(vec: Vector3): auto = obj

  result = getAst makeProc(op, obj)
 
let vec = Vector3(x: 1, y: 2, z: 3)

echo vec.zy # prints '(x: 3.0, y: 2.0)'
echo vec.yzx # prints '(x: 2.0, y: 3.0, z: 1.0)'
```
A common usecase for doing shader- and graphics-related math.

**Object Proxying**
```nim
type Proxy[T] = object
  obj: T

template `=call`(proxy: Proxy[T], op: untyped): untyped =
  # intercept field access and/or routine calls
  ...
  proc op(proxy: Proxy[T]): auto = op(proxy.obj)

let proxy = Proxy[string](obj: "Hello")

echo $proxy # prints 'Hello'
```
This usecase could lessen the need for [converters](https://nim-lang.org/docs/manual.html#converters), since most operations can be forwarded through a proxy to the underlying object. This could also allow things like mapping field access to entity component access in an ECS.

**Remote function invocation:**
```nim
type Network = object

template `=call`(obj: Network, op: untyped) =
  # perform operation to execute remote function or API
  ...

doStuffOverNetwork(Network())
```
This could be used to access arbitrary columns in a remote database for example, or execute an API with a nice syntax.

## Comparison with Existing Solutions

* The experimental [`dotOperator`](https://nim-lang.org/docs/manual_experimental.html#special-operators-dot-operators) facilitates many of the usecases in this RFC. Some disadvantages of this approach are:
  * The meaning of `.` is overloaded:
    * default: object field access
    * default: sugar for routine invocation
    * `dotOperator`: arbitrary code can be generated that may or may not result in field access or a routine invocation.
  * `.` becomes special and can do things that UFCS and command invocation syntax cannot in the context of routine invocation.
  * Allows a limited number of operator overloads. For example, `.=` can be overloaded but `.+=` cannot.
  * Seems to be emulating ternary operators (e.g. `.=` and `.()`), which don't exist in Nim. This leads to a lot of special casing and perhaps complexity in the compiler.

* [Term-rewriting macros](https://nim-lang.org/docs/manual_experimental.html#term-rewriting-macros) could be modified to provide similar or even greater flexibility as mentioned in [this comment](https://github.com/nim-lang/RFCs/issues/341#issuecomment-778572488). However it is probably not an ideal solution, since TRMs weren't designed for these usecases.

## Proposed Mechanics

* The compiler will attempt to call `=call` when a call to an undefined routine is encountered, and then will try to call the routine.

```nim
proxy.test = 1 # `test=` is an undeclared routine
# If `=call` is defined for `typeof(proxy)`, the above effectively expands to:
# `=call`(proxy, `test=`, 1)
# `test=`(proxy, 1)
```

* The compiler will greedily consider any operators after the undeclared routine call as part of the routine name passed to `=call`.

```nim
proxy.test += 1 # `test` is an undeclared routine
# If `=call` is defined for `typeof(proxy)`, the above effectively expands to:
# `=call`(proxy, `test+=`, 1)
# `test+=`(proxy, 1)
```

## Conflicts with Current Behaviour

* Dot(-like) operator overloading conflicts with the following usage:
```nim
discard proxy.?test # Error: undeclared identifier: '.?'
```
This feature could be desirable if the programmer wishes to prefix dot operator access with a special operator to signal "here be dragons".

The solution to this could be related to disallowing the overloading of dot-like operators altogether, and can be explored in a different RFC.

## Related Literature

Dlang implements a [similar concept](https://dlang.org/spec/operatoroverloading.html) for its operators, but instead of passing an AST to its "type-bound templates", it passes a compile-time string containing the invoked operator. It is then up to the programmer to define and implement functionality for each possible operator passed to these templates.

## Links

* Sample playground: https://play.nim-lang.org/#ix=2Ttc
* Forum post: https://forum.nim-lang.org/t/7752
* Dot operators: https://nim-lang.org/docs/manual_experimental.html#special-operators-dot-operators
* Converters: https://nim-lang.org/docs/manual.html#converters
* Related RFC: https://github.com/nim-lang/RFCs/issues/341
* Dlang's operator overloading: https://dlang.org/spec/operatoroverloading.html

**Comment by Varriount** [comment]
So, just to confirm my interpretation:
  - Currently, dot operators can produce any result, possibly breaking the semantic expectation the `.` syntax has.
  - The proposal is to essentially refine these operators so that they are expected to produce a routine (or is it only procedures?) of some sort, and have the compiler produce code that calls the routine.

**Comment by n0bra1n3r** [comment]
@Varriount Yes for both points.

The proposal also goes a bit further than your second point though. It tries to do away with special casing of `.` (or any other operator), and instead allows the programmer to generate a procedure if it is not declared when a call to it is encountered by the compiler. I tried to describe this under the Proposed Mechanics section a bit.

Please let me know if anything is not clear. I really want to make this a good RFC. Thanks!

**Comment by Araq** [comment]
The RFC is well-written but there is a fundamental design tension between "let's have custom dot-like operators" and "the behavior of the dot notation is overridable". I much prefer "custom dot-like operators" which rules out "Object Proxying" entirely, no matter the details of how it's done.

Having said that, a design should probably focus on allowing convenient user definable smart pointers.

**Comment by n0bra1n3r** [comment]
@Araq I see. So to clarify, object proxying is something you want to only be possible in specific scenarios (like for implementing smart pointers)? Or is it that object proxying is evil, full stop?

I guess my gripe with dot operator overloading is that it gives `.` (or any variation of it) changes the meaning of `.` from "just another way to call functions or to access object fields". I'm not sure this RFC completely solves that either though, but at least the result of any `.` operation is guaranteed to be the result of a function call (or field access).

**Comment by Varriount** [comment]
> The RFC is well-written but there is a fundamental design tension between "let's have custom dot-like operators" and "the behavior of the dot notation is overridable".

Another way to look at this might be: Term-rewriting macros are either not powerful enough, or not usable enough (from an ease-of use perspective) to apply to the expressions currently targeted by dot operators.

**Comment by Araq** [comment]
> I guess my gripe with dot operator overloading is that it changes the meaning of . (or any variation of it) from "just another way to call functions or to access object fields" to "a special way to generate arbitrary code, plus the other stuff".

That's a good way to put it, here is another one: Turning `a.b` into `a["b"]` (dynamic access that can fail) is what I dislike most -- you get most of the problems of dynamic typing within Nim.

**Comment by n0bra1n3r** [comment]
> you get most of the problems of dynamic typing within Nim.

Yes. I would dare say that any operator that has `.` in it (`.?`, `.!`, etc.) like what is proposed in RFC https://github.com/nim-lang/RFCs/issues/341 has this weakness. This RFC also carries some of those disadvantages, but I would go so far as to say that this one is superior to dot-like operators because of the limitations it imposes, as well as the flexibility it allows.

One more important advantage of this RFC (aside from what's already mentioned) over dot/dot-like operators is that operators can be declared in any module, separate from the type they operate on. `=call` in this RFC however would behave similarly to `=destroy`, which has to be declared in the same module as the type it operates on. This makes it clear that there is a certain amount of magic involved when working with a type that has `=call`.

**Comment by Araq** [comment]
`=call` is an alien beast though, the other type bound operators are lifted automtically, you define `=` for CustomObj and it's not skipped for a tuple of CustomObj, `=call` has no such lifting requirements.

**Comment by n0bra1n3r** [comment]
Well, can't argue against that, sounds like a showstopper... 😂

**Comment by Araq** [comment]
It can easily be implemented either way.
